django/db/models/expressions.py

```diff
@@ -360,27 +360,30 @@ def flatten(self):
             if expr:
                 yield from expr.flatten()
 
+    @property
+    def identity(self):
+        identity = (self.__class__,)
+        output_field = self.__dict__.get('output_field')
+        if isinstance(output_field, fields.Field):
+            # Field instances need to be special cased because their equality
+            # is determined by their creation_counter.
+            name, path, args, kwargs = output_field.deconstruct()
+            args = tuple(args)
+            kwargs = tuple([
+                (key, tuple(value) if isinstance(value, list) else value)
+                for key, value in kwargs.items()
```

Comment:
I don't think we can rely on `kwargs.items()` to be ordered. I suggest using `sorted` to enforce an ordering.

Reply:
@Ian-Foote not sure how much if this should be considered an implementation detail but dicts with the same keys will necessarily yield items in the same order since they share the same `PYTHONHASHSEED`. I concede that the ordering of `items()` will be unstable between interpreter launches but this identity is not required to be.

Anyway, I don't mind wrapping `kwargs.items()` in `sorted` in all cases but it might be unnecessary.

Reply:
Insertion order is maintained from Python 3.6 onwards so sorting is only required for 3.5:

https://www.python.org/dev/peps/pep-0468/

Reply:
@pope1ni I get that Python 3.6 dicts maintain ordering. My point is that under Python < 3.6 dict with the same keys will be ordered in the same manner for a shared `PYTHONHASHSEED` independently of their insertion order.

e.g.

```python
>>> list({key: 'value' for key in range(5000)}) == list({key: 'value' for key in range(5000)})
True
>>> list({'foo': 'bar', 'bar': 'foo'}) == list({'bar': 'foo', 'foo': 'bar'})
True # False on Python 3.6+
>>> a, b = {'foo': 'bar'}, {'bar': 'foo'}
>>> a['bar'] = 'foo'
>>> b['foo'] = 'bar'
>>> list(a) == list(b)
True # False on Python 3.6+
```

The same is true for sets on all versions of Python by the way.
```python
>>> list({'a', 'b'}) == list({'b', 'a'})
True # Even on Python 3.6+
```

Since `identity` is only used for _identification_ purpose we could have relied on this property to completely avoid sorting items because as soon as there is a key mismatch the tuple would be in a possibly totally different order. In other words, `sorted` is only required on Python 3.6 to support non deterministic insertion order of the kwargs returned by `Field.deconstruct` which is unlikely. Try writing a `deconstruct` function that breaks the above code on Python 3.6+ and you'll see for yourself. On Python < 3.6 it's even harder as you have to return an `OrderedDict` and make sure to initialize it in a different order for the same signature.

Anyway, at this point I'll just use `sorted` to avoid confusing future readers.

Reply:
Sorry, I wasn't particularly clear - I was referring to keyword arguments, not dicts in general as highlighted by your examples. Was just highlighting the point as the preceding comments seemed to only discuss behaviour for Python < 3.5

Not that it matters... Using `sorted` makes sense for now.